server/src/main/java/org/elasticsearch/common/io/stream/RecyclerBytesStreamOutput.java

```diff
@@ -77,36 +94,90 @@ public void writeBytes(byte[] b, int offset, int length) {
 
         Objects.checkFromIndexSize(offset, length, b.length);
 
-        // get enough pages for new size
-        final int pageSize = this.pageSize;
-        int currentPageOffset = this.currentPageOffset;
+        final int currentPageOffset = this.currentPageOffset;
         if (length > pageSize - currentPageOffset) {
-            ensureCapacity(length);
-            currentPageOffset = this.currentPageOffset;
+            writeMultiplePages(b, offset, length);
+        } else {
+            final BytesRef currentPage = currentBytesRef;
+            final byte[] bytes = currentPage.bytes;
+            final int destOffset = currentPage.offset + currentPageOffset;
+            System.arraycopy(b, offset, bytes, destOffset, length);
+            this.currentPageOffset = currentPageOffset + length;
         }
+    }
+
+    private void writeMultiplePages(byte[] b, int offset, int length) {
+        ensureCapacity(length);
 
-        // bulk copy
+        BytesRef currentPage = currentBytesRef;
+        int currentPageOffset = this.currentPageOffset;
         int bytesToCopy = length;
         int srcOff = offset;
-        int j = 0;
         while (true) {
-            BytesRef currentPage = pages.get(pageIndex + j).v();
-            int toCopyThisLoop = Math.min(pageSize - currentPageOffset, bytesToCopy);
-            System.arraycopy(b, srcOff, currentPage.bytes, currentPage.offset + currentPageOffset, toCopyThisLoop);
+            final int toCopyThisLoop = Math.min(pageSize - currentPageOffset, bytesToCopy);
+            final int destOffset = currentPage.offset + currentPageOffset;
+            System.arraycopy(b, srcOff, currentPage.bytes, destOffset, toCopyThisLoop);
             srcOff += toCopyThisLoop;
             bytesToCopy -= toCopyThisLoop;
             if (bytesToCopy > 0) {
                 currentPageOffset = 0;
+                currentPage = pages.get(++pageIndex).v();
             } else {
                 currentPageOffset += toCopyThisLoop;
                 break;
             }
-            j++;
         }
         this.currentPageOffset = currentPageOffset;
+        this.currentBytesRef = currentPage;
+    }
+
+    @Override
+    public void writeVInt(int i) throws IOException {
```

Comment:
vInt is very commonly written (header for bytes reference, map item count, etc, etc). Since it is pretty expensive to write we create a specific variant that works for the recycler stream variant.